libfwupdplugin/fu-fmap-firmware.c

```diff
@@ -205,26 +90,25 @@ fu_fmap_firmware_parse (FuFirmware *firmware,
 
 	/* only search for the fmap signature if not fuzzing */
 	if ((flags & FWUPD_INSTALL_FLAG_NO_SEARCH) == 0) {
-		if (!fmap_find (buf, bufsz, &offset, error)) {
-			g_prefix_error (error, "cannot find fmap in image: ");
+		if (!fu_fmap_firmware_find_offset (self, buf, bufsz, error))
 			return FALSE;
-		}
 	}
 
 	/* load header */
+	offset = priv->offset;
 	if (!fu_memcpy_safe ((guint8 *) &fmap, sizeof(fmap), 0x0,	/* dst */
-			     buf, bufsz, offset,			/* src */
+			     buf, bufsz, priv->offset,			/* src */
 			     sizeof(fmap), error))
 		return FALSE;
 	priv->base = GUINT64_FROM_LE (fmap.base);
 
-	if (GUINT32_FROM_LE (fmap.size) != bufsz) {
+	if (GUINT32_FROM_LE (fmap.size) != bufsz - priv->offset) {
```

Comment:
Not sure this is correct...

This error pops on this version of the change:
`file size incorrect, expected 0x80000 got 0x4fa80`

Reply:
Can you send me a link of an image that's valid and that FuFmapFirmware fails to parse please? I've added one test, and we probably need a few more.

Reply:
Also; you sure you tested the latest version? I got that too on the interim version of the commit that I repushed.

Reply:
I'm at commit 40a3bb786916b93f79339de5c0db5782ac3cdd21.

This binary has the issue too. 

https://fwupd.org/downloads/b67bda9d75094f7d534e706557ce0a5ff973e9a3543d39eb1ba78b833ec410a6-servo_micro_v2.4.17.cab

Reply:
I can't share with you the .bin I tested before, but the servo_micro_v2.4.17 has a similar issue. Here's the exact error from the binary I just linked to:

`file size incorrect, expected 0x20000 got 0x165c0`

Reply:
Hmm, so there's two bugs. Is the FuFmapArea.offset supposed to be relative to the start of the .bin or relative to the offset that __FMAP__ is specified at? The .bin would suggest the former. The FuFmap.size also seems to specify the file size, not the size of the __FMAT__ section...

Reply:
https://www.chromium.org/chromium-os/firmware-porting-guide/fmap

yeah, fmap_size is the full size of the file.

Reply:
Area.offsets are from the start of the .bin.

Reply:
Okay, the issue was in the way we wrote the FMAP file; I've changed the reading back to what we had before and than can now pass it's new self tests and also parse the file we know should work. I'll re-push when we've worked out the -private thing.

Reply:
@bleungatchromium looks good to you now?